File: src/main/java/com/raychenon/leetcode/CustomSortString.java
package com.raychenon.leetcode;

import java.util.HashMap;
import java.util.Map;

/**
 * User: raychenon
 * Date: 2019-11-06
 * https://leetcode.com/problems/custom-sort-string/
 */
public class CustomSortString {

    /**
     * Time complexity: O(S.length + T.length)
     * Space complexity: O(T.length)
     *
     * @param S
     * @param T
     * @return
     */
    public String customSortString(String S, String T) {
        Map<Character, Integer> map = new HashMap<>();

        // create a map frequency of chars for T
        for (Character c : T.toCharArray()) {
            map.put(c, map.getOrDefault(c, 0) + 1);
        }

        StringBuilder str = new StringBuilder();

        // append all chars in S
        for (Character c : S.toCharArray()) {
            if (map.containsKey(c)) {
                for (int i = 0; i < map.getOrDefault(c, 0); i++) {
                    str.append(c);
                }
            }
            map.remove(c);
        }

        // append the remaining chars in T
        // How keys in Map are "magically" sorted in right alphabetical order ?
        for (Map.Entry<Character, Integer> entry : map.entrySet()) {
            for (int i = 0; i < entry.getValue(); i++) {
                str.append(entry.getKey());
            }
        }

        return str.toString();

    }
}

File: src/main/kotlin/com/raychenon/kotlin/leetcode/CustomSortString.kt
package com.raychenon.kotlin.leetcode

/**
 * User: raychenon
 * Date: 2019-11-06
 * https://leetcode.com/problems/custom-sort-string/
 */
object CustomSortString {

    /**
     * Time complexity: O(S.length + T.length)
     * Space complexity: O(T.length)
     * @param S
     * @param T
     * @return
     */
    fun customSortString(S: String, T: String): String {
        var map = mutableMapOf<Char, Int>()
        // create a map of occurrences of chars for T
        for (c in T) {
            map.put(c, map.getOrDefault(c, 0) + 1)
        }

        // add chars from S
        var str = StringBuilder()
        for (c in S) {
            if (map.contains(c)) {
                for (i in 1..map.getOrDefault(c, 0)) {
                    str.append(c)
                }
                map.remove(c)
            }
        }

        // append the remaining chars in T
        // How keys in Map are "magically" sorted in right alphabetical order ?
        for ((k, v) in map) {
            for (i in 1..v) {
                str.append(k)
            }
        }

        return str.toString()
    }

}